Apply new conventions on Turek Hron FSI3 case (#130)
precice/tutorials: turek-hron-fsi3/README.md

---
title: Turek Hron FSI3
permalink: tutorials-turek-hron-fsi3.html
keywords: OpenFOAM, deal.II, verification
summary: The Turek-Hron FSI cases are well-established numerical benchmarks and, therefore, well suited for verification of preCICE itself and the used adapters. In this tutorial, we focus on the FSI3 case, which presents the most challenging case in terms of added mass. Please note that the meshes of this case are significantly finer than for other tutorials. Running the simulation might take a few hours. We do not recommend to run this tutorials as your first preCICE tutorial.  
---

## Setup

The setup is shown schematically here:

> ![FSI3 setup](images/tutorials-turek-hron-fsi3-setup.png)

Reviewer: These are the results from the previous cylinderFlap case. I guess we need a scheme with just the setup, or an image of the actual results.

Reviewer 2: I can probably provide some more recent.

Author: That would be good. If we show physical results they should coincide 100% with the case. Alternative would be to only show a setup image. Please directly push to the source branch.

For more information please refer to the original publication of the benchmark [1].

## Available solvers

> hier auch wo mann sie herbekommt

Reviewer: TODO (but do we want this here?)

Author: That was just an Easter egg :smile:

Fluid participant:

* OpenFOAM. For more information, have a look at the [OpenFOAM adapter documentation](adapter-openfoam-overview.html). 

> {% include important.html content="For the parabolic inflow profile, this tutorial requires groovyBC. groovyBC is part of swak4Foam, which can be cloned from [this repository](https://github.com/Unofficial-Extend-Project-Mirror/openfoam-extend-swak4Foam-dev.git). Please follow the building instructions there." %}

Reviewer: Bernhard Gscheider (the maintainer of swak4Foam) recommends to not use "the Git repository that is going around". I am not sure why and I cannot find the official repository at the moment (which should be a Mercurial repository on BitBucket). I guess the most official resource for now is https://openfoamwiki.net/index.php/Contrib/swak4Foam (he also maintains the OpenFOAMWiki, but apparently has not updated this page for a while, I pinged him).

Reviewer 2: 2010 :sweat_smile: 
Anyway this might be interesting. Our whole FSI validation is based on the unofficial git mirror. I hope there are not that big differences for GroovyBC

Reviewer: More information: https://twitter.com/swakPyFoam/status/1338566289913163777
Let's use both links and clarify that we used the Git repository, but it is unofficial.

Author: I added both links. We should add that we used the Git repo in the paper, not here. Anyway, I guess not too much can go wrong with a parabolic inflow profile.

Solid participant:

> * deal.II. For more information, have a look at the [deal.II adapter documentation](adapter-dealii-overview.html). This tutorial requires the nonlinear solid solver. Please copy the nonlinear solver executable to this folder or make it discoverable at runtime.

Reviewer: ```suggestion
* deal.II. For more information, have a look at the [deal.II adapter documentation](adapter-dealii-overview.html). This tutorial requires the nonlinear solid solver. Please copy the nonlinear solver executable to this folder or make it discoverable at runtime and update the `solid-dealii/run.sh` script.
```

## Running the Simulation

Open two separate terminals and start each participant by calling the respective run script.

```
cd fluid-openfoam
./run.sh
```
and
```
cd solid-dealii
./run.sh
```

You can also run OpenFOAM in parallel by `./run.sh -parallel`.

> Moreover, the name of your solver might differ, depending on your OpenFOAM version. Have a look in the `fluid-openfoam/system/controlDict` file and set the appropriate solver name.

Reviewer: ```suggestion
For older OpenFOAM versions, the solver name will differ. If you are using OpenFOAM v1712 / 5.x or older have a look in the `fluid-openfoam/system/controlDict` file and set the appropriate solver name.
```

> There is an [open issue](https://github.com/precice/openfoam-adapter/issues/26) that leads to additional "empty" result directories when running with some OpenFOAM versions, leading to inconveniences during post-processing. Please run the script `removeObsoleteSolvers.sh` to delete the additional files.

Reviewer: ```suggestion
There is an [known issue](https://github.com/precice/openfoam-adapter/issues/26) that leads to additional "empty" result directories when running with some OpenFOAM versions, leading to inconveniences during post-processing. Please run the script `removeObsoleteSolvers.sh` to delete the additional files before importing your results in ParaView.
```

Reviewer: Let's also move this to the post-processing part.

Author: done

> You may adjust the end time in the precice-config_*.xml, or interupt the execution earlier if you want.

Reviewer: ```suggestion
You may adjust the end time in the `precice-config_*.xml`, or interupt the execution earlier if you want.
```

Author: I also removed the `_*`.

> In the first few timesteps, many coupling iterations are required for convergence. Don't loose hope, things get better quickly. 

Reviewer: ```suggestion
In the first few timesteps, many coupling iterations are required for convergence. Don't lose hope, things get better quickly. 
```

## Postprocessing
   

> You can visualize the results of the coupled simulation using e.g. ParaView. Fluid results are in the OpenFOAM format and you may load the `Fluid.foam` file. Solid results are in VTK format. If you want to visualize both domains with ParaView, keep in mind that the deal.II solver writes results every few timesteps, while the OpenFOAM solver writes in reference to simulated time. For this reason, make sure that you use compatible write intervals. You may also need to convert the OpenFOAM results to VTK (with the command `foamToVTK`). 
> Moreover, we defined a watchpoint at the flap tip and can observe the perpendicular displacement with the script `plotDisplacement.sh`. 

Reviewer: Let's split this in three paragraphs, as things get complicated:

```suggestion
You can visualize the results of the coupled simulation using e.g. ParaView. Fluid results are in the OpenFOAM format and you may load the `Fluid.foam` file. Solid results are in VTK format.

If you want to visualize both domains with ParaView, keep in mind that the deal.II solver writes results every few timesteps, while the OpenFOAM solver writes in reference to simulated time. For this reason, make sure that you use compatible write intervals. You may also need to convert the OpenFOAM results to VTK (with the command `foamToVTK`). 

Moreover, as we defined a watchpoint at the flap tip (see `precice-config.xml`), we can plot it with gnuplot using the script `plotDisplacement.sh`.  The resulting graph shows the vertical (y) displacement of the tip of the flap.
```

(I hope that `y` is correct here)

> TODO: update picture according to actual used mesh.

Reviewer: TODO

Reviewer 2: I recently noticed that the foamToVTK only helps for `.org` versions. For `.com` the problem remains.

Author: Which problem are you referring to exactly? The non-synchronized files? Could you make a text suggestion? (or directly push to the source branch).

Reviewer 2: Yes, the time sync. Since OpenFOAM shows the exact motion of the structure, all in all it is not such a big problem. I have not yet a proper solution for this problem, it might be helpful if you could upload a synchronized nutils result file in https://github.com/precice/dealii-adapter/issues/33

Author: OK, but I would exclude the time synch problem from this PR. I will add files in the dealii issue. TODO here is done.

You can visualize the results of the coupled simulation using e.g. ParaView. Fluid results are in the OpenFOAM format and you may load the `Fluid.foam` file. Solid results are in VTK format.

If you want to visualize both domains with ParaView, keep in mind that the deal.II solver writes results every few timesteps, while the OpenFOAM solver writes in reference to simulated time. For this reason, make sure that you use compatible write intervals. You may also need to convert the OpenFOAM results to VTK (with the command `foamToVTK`).

There is an [known issue](https://github.com/precice/openfoam-adapter/issues/26) that leads to additional "empty" result directories when running with some OpenFOAM versions, leading to inconveniences during post-processing. Please run the script `removeObsoleteSolvers.sh` to delete the additional files before importing your results in ParaView.

Moreover, as we defined a watchpoint at the flap tip (see `precice-config.xml`), we can plot it with gnuplot using the script `plotDisplacement.sh`.  The resulting graph shows the vertical (y) displacement of the tip of the flap.

![FSI3 watchpoint](images/tutorials-turek-hron-tip-plot.png)

> 

Reviewer: ```suggestion

Before running the simulation again, you may want to cleanup any result files using the script `Allclean`.
```

## Mesh refinement

> In `fluid-openfoam/system/`, we provide three different fluid meshes:

Reviewer: How is the user supposed to select a different mesh?
